src/ApparentHorizons/SpherepackIterator.hpp

```diff
@@ -0,0 +1,116 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+///
+/// \file
+/// Defines class SpherepackIterator.
+
+#pragma once
+
+#include <ostream>
+#include <vector>
+
+#include "ErrorHandling/Assert.hpp"
+
+/// \ingroup Spectral
+/// \brief
+/// Iterates over spectral coefficients stored in SPHEREPACK format.
+///
+/// \details
+/// The internal SPHEREPACK ordering is not intuitive, so
+/// SpherepackIterator exists for the purpose of iterating over the
+/// array and determining which specific values of l and m correspond
+/// to each iteration.
+/// \example
+/// \snippet Test_SpherepackIterator.cpp spherepack_iterator_example
+class SpherepackIterator {
+ public:
+  /// 'a' and 'b' are what SPHEREPACK calls the two FORTRAN arrays.
+  /// 'a' holds real and 'b' holds the imaginary parts of the coefficients.
+  enum class CoefficientArray { a, b };
+  SpherepackIterator(size_t l_max_input, size_t m_max_input, size_t stride = 1);
+
+  size_t l_max() const noexcept { return l_max_; }
+  size_t m_max() const noexcept { return m_max_; }
+  size_t n_th() const noexcept { return n_th_; }
+  size_t n_ph() const noexcept { return n_ph_; }
+  size_t stride() const noexcept { return stride_; }
+
+  /// Size of the combined a/b array *in the SPHEREPACK dimensions*.
+  /// (so the actual size of the array is 'stride' times larger).
+  size_t spherepack_array_size() const noexcept {
+    return (l_max_ + 1) * (m_max_ + 1) * 2;
+  }
+
+  SpherepackIterator& operator++() noexcept {
+    ASSERT(current_compact_index_ != offset_into_spherepack_array.size(),
+           "Incrementing an invalid iterator: "
+               << current_compact_index_ << " "
+               << offset_into_spherepack_array.size());
+    ++current_compact_index_;
+    return *this;
+  }
+
+  explicit operator bool() const noexcept {
+    return current_compact_index_ < offset_into_spherepack_array.size();
+  }
+
+  /// Current index into combined a/b array storage. Used to index a
+  /// vector representing spectral coefficients stored in SPHEREPACK
+  /// format.
+  size_t operator()() const noexcept {
+    return offset_into_spherepack_array[current_compact_index_];
+  }
+
+  /// current values of l and m, and of which array (real or imaginary parts)
+  size_t l() const noexcept { return compact_l_[current_compact_index_]; }
+  size_t m() const noexcept { return compact_m_[current_compact_index_]; }
```

Comment:
I would add a separate Doxygen comment for this guy,
`Whether the iterator points to an element of 'a' or 'b', i.e. points to the real or imaginary part of a spectral coefficient`